File: tests/solutions_cpp/clone_graph.cpp
#include "json_helper.h"

class Solution {
public:
    GraphNode* cloneGraph(GraphNode* node) {
        if (!node) return nullptr;
        unordered_map<GraphNode*, GraphNode*> m;
        return clone(node, m);
    }
private:
    GraphNode* clone(GraphNode* node, unordered_map<GraphNode*, GraphNode*>& m) {
        auto it = m.find(node);
        if (it != m.end()) return it->second;
        GraphNode* copy = new GraphNode(node->val);
        m[node] = copy;
        for (auto* nb : node->neighbors) {
            copy->neighbors.push_back(clone(nb, m));
        }
        return copy;
    }
};

int main() {
    auto args = parse_args();
    Solution sol;
    auto g = build_graph(args[0]);
    cout << graph_to_json(sol.cloneGraph(g)) << endl;
    return 0;
}

File: tests/solutions_js/clone_graph.js
class Solution {
    cloneGraph(node) {
        if (!node) return null;

        const visited = new Map();

        const dfs = (n) => {
            if (visited.has(n)) return visited.get(n);
            const clone = new Node(n.val);
            visited.set(n, clone);
            for (const neighbor of n.neighbors) {
                clone.neighbors.push(dfs(neighbor));
            }
            return clone;
        };

        return dfs(node);
    }
}
